### PR TITLE
Simplifying order handling by moving expiration/modification to traders

### PR DESCRIPTION
**Pull request by stevenewald**
Now that we store direct references to traders (instead of string IDs, which then have to be looked up in the trader manager), we can directly modify them from orders.

This is better design because trader handling of order expiration/modification and addition is in the domain of the traders - not the bot container or engine.

### CODE REVIEW

**Comment by stevenewald** [comment]
This introduces `StoredMatch`, which is partially (and will be fully with the new engine) the way the exchange stores matches after they're created. This allows us to store direct references to traders involved in the match (making operations based on these matches much easier). They're converted into the RMQ-transmissible format Match before being sent out

File: exchange/src/exchange/bots/bot_types/retail.cpp
Comment (egelja) [inline exchange/src/exchange/bots/bot_types/retail.cpp:1]: these constants should be in `config.h`

File: exchange/src/exchange/rabbitmq/order_handler/RabbitMQOrderHandler.cpp [hunk 1]
```diff
@@ -33,13 +32,21 @@ RabbitMQOrderHandler::handle_incoming_market_order(
...
+    std::vector<Match> rmq_matches;
+    rmq_matches.reserve(matches.size());
     for (const auto& match : matches) {
-        std::string buyer_id = match.buyer_id;
-        std::string seller_id = match.seller_id;
+        std::string buyer_id = match.buyer->get_id();
+        std::string seller_id = match.seller->get_id();
+        Match b_match{match.ticker,   match.side, match.price,
+                      match.quantity, buyer_id,   seller_id};
+        rmq_matches.push_back(b_match);
```
Comment (egelja) [inline exchange/src/exchange/rabbitmq/order_handler/RabbitMQOrderHandler.cpp:42]:
```suggestion
        std::string buyer_id = match.buyer->get_id();
        std::string seller_id = match.seller->get_id();
        Match b_match{match.ticker,   match.side, match.price,
                      match.quantity, std::move(buyer_id),   std::move(seller_id)};
        rmq_matches.push_back(std::move(b_match));
```

File: exchange/src/exchange/tickers/engine/engine.cpp [hunk 2]
```diff
@@ -205,10 +204,10 @@ Engine::attempt_matches_( // NOLINT (cognitive-complexity-*)
...
         sell_order.quantity -= quantity_to_match;
 
-        if (sell_order.trader->get_type() != manager::BOT
-            || buy_order.trader->get_type() != manager::BOT) {
-            logger.log_event(to_match);
-        }
+        // if (sell_order.trader->get_type() != manager::BOT
+        // || buy_order.trader->get_type() != manager::BOT) {
+        // logger.log_event(to_match);
+        // }
```
Comment (egelja) [inline exchange/src/exchange/tickers/engine/engine.cpp:210]: delete pls

File: exchange/src/exchange/tickers/engine/new_engine.hpp [hunk 3]
```diff
@@ -85,7 +92,7 @@ class NewEngine {
         return matches;
     }
 
-    uint64_t current_tick_;
+    uint64_t current_tick_ = 0;
```
Comment (egelja) [inline exchange/src/exchange/tickers/engine/new_engine.hpp:95]: top of class pls

File: exchange/src/exchange/tickers/manager/ticker_manager.hpp [hunk 4]
```diff
@@ -34,28 +35,35 @@ class EngineManager : public nutc::ticks::TickObserver {
             auto [removed, added, matched] =
                 engine.on_tick(new_tick, ORDER_EXPIRATION_TIME);
 
-            for (matching::StoredOrder& order : added) {
-                if (order.trader->get_type() != manager::TraderType::BOT)
-                    continue;
-                bot_containers_.at(order.ticker)
-                    .process_order_add(
-                        order.trader->get_id(), order.side, order.price * order.quantity
+            std::for_each(
```
Comment (egelja) [inline exchange/src/exchange/tickers/manager/ticker_manager.hpp:38]: use a for loop

for_each is slower

File: exchange/src/exchange/tickers/manager/ticker_manager.hpp [hunk 5]
```diff
@@ -34,28 +35,35 @@ class EngineManager : public nutc::ticks::TickObserver {
...
+                }
+            );
 
-            for (matching::StoredOrder& order : removed) {
-                if (order.trader->get_type() != manager::TraderType::BOT)
-                    continue;
-                bot_containers_.at(order.ticker)
-                    .process_order_expiration(
-                        order.trader->get_id(), order.side, order.price * order.quantity
+            std::for_each(
```
Comment (egelja) [inline exchange/src/exchange/tickers/manager/ticker_manager.hpp:47]: use a for loop

for_each is slower

File: exchange/src/exchange/tickers/manager/ticker_manager.hpp [hunk 6]
```diff
@@ -34,28 +35,35 @@ class EngineManager : public nutc::ticks::TickObserver {
...
                     );
-            }
+                }
+            );
 
-            for (Match& order : matched) {
-                // TODO(stevenewald): check if bot beforehand?
-                bot_containers_.at(ticker).process_order_match(order);
-            }
+            std::for_each(
```
Comment (egelja) [inline exchange/src/exchange/tickers/manager/ticker_manager.hpp:56]: use a for loop

for_each is slower

File: exchange/src/exchange/traders/trader_types/bot_trader.hpp [hunk 7]
```diff
@@ -119,6 +107,44 @@ class BotTrader : public manager::GenericTrader {
         return (get_long_interest() + get_short_interest()) / get_interest_limit();
     }
 
+    void
+    process_order_expiration(
```
Comment (egelja) [inline exchange/src/exchange/traders/trader_types/bot_trader.hpp:111]: should be in cpp file

File: exchange/src/exchange/traders/trader_types/bot_trader.hpp [hunk 8]
```diff
@@ -119,6 +107,44 @@ class BotTrader : public manager::GenericTrader {
...
+            modify_open_bids(-1);
+        }
+        else {
+            modify_short_capital(-total_cap);
+            modify_open_asks(-1);
+        }
+    }
+
+    void
+    process_order_add(
```
Comment (egelja) [inline exchange/src/exchange/traders/trader_types/bot_trader.hpp:130]: should be in cpp file

File: exchange/src/exchange/traders/trader_types/generic_trader.hpp [hunk 9]
```diff
@@ -81,6 +83,34 @@ class GenericTrader {
...
+    process_order_expiration(const std::string&, messages::SIDE, double, double)
+    {}
+
+    // ticker, price, side, quantity
+    virtual void
+    process_order_match(
+        const std::string& ticker, messages::SIDE side, double price, double quantity
+    )
+    {
+        double total_cap = price * quantity;
```
Comment (egelja) [inline exchange/src/exchange/traders/trader_types/generic_trader.hpp:103]: should be in cpp file

File: exchange/test/src/test_utils/macros.cpp [hunk 10]
```diff
@@ -4,6 +4,20 @@
...
+StoredOrder
+make_stored_order(MarketOrder& order, const manager::ClientManager& manager)
+{
+    return StoredOrder{
+        manager.get_trader(order.client_id),
+        order.side,
+        order.ticker,
+        order.quantity,
+        order.price,
+        0
```
Comment (egelja) [inline exchange/test/src/test_utils/macros.cpp:17]: add a comment detailing what this 0 means

File: exchange/test/src/unit/logging/logging_orders.cpp [hunk 11]
```diff
@@ -53,5 +53,5 @@ TEST_F(UnitLoggingOrders, LogMatches)
     auto [matches2, ob_updates2] = add_to_engine_(order2);
 
     auto& logger = Logger::get_logger();
-    EXPECT_NO_FATAL_FAILURE(logger.log_event(matches2.at(0)));
+    // EXPECT_NO_FATAL_FAILURE(logger.log_event(matches2.at(0)));
```
Comment (egelja) [inline exchange/test/src/unit/logging/logging_orders.cpp:56]: fix this

**Review by egelja — CHANGES_REQUESTED** [review]
don't use `std::for_each`.  otherwise looks good

**Pull request merged**